File: src/vlearning/umpire/VUmpire.java
package vlearning.umpire;

import java.util.Arrays;

import vlearning.agent.*;
import common.*;
import qlearning.environment.TicTacToe;
import vlearning.environment.*;

/**
 * The umpire.<br>
 * Project: Reinforcement learning.<br>
 * 
 * @author Fan Min<br>
 *         www.fansmale.com, https://github.com/FanSmale/MFReinforcement.<br>
 *         Email: minfan@swpu.edu.cn, minfanphd@163.com.<br>
 *         Date Created: August 16, 2020.<br>
 *         Last modified: August 20, 2020.
 * @version 1.0
 */

public class VUmpire {
	/**
	 * The environment for the agents.
	 */
	VTicTacToe environment;

	/**
	 * Two players.
	 */
	VAgent[] agentArray;

	/**
	 * The win times of agents. 0 stands for TIE.
	 */
	int[] winTimesArray;

	/**
	 * The winner.
	 */
	int winner;

	/**
	 * The checkerboard status.
	 */
	int[][] playCheckerboard;

	/**
	 ****************** 
	 * The constructor.
	 * 
	 * @param paraEnvironment
	 *            The environment.
	 *            @param paraAgentArray The agents.
	 ****************** 
	 */
	public VUmpire() {
		environment = new VTicTacToe();
		agentArray = new VAgent[2];
		for (int i = 0; i < agentArray.length; i++) {
			agentArray[i] = new VAgent(environment, i + 1);
		}//Of  for i
		
		playCheckerboard = new int[VTicTacToe.SIZE][VTicTacToe.SIZE];
		
		winTimesArray = new int[3];
		resetForGame();
	}// Of the constructor

	/**
	 ****************** 
	 * Reset for the new game.
	 ****************** 
	 */
	public void resetForGame() {
		winner = VTicTacToe.EMPTY;
		for (int i = 0; i < VTicTacToe.SIZE; i++) {
			for (int j = 0; j < VTicTacToe.SIZE; j++) {
				playCheckerboard[i][j] = VTicTacToe.EMPTY;
			}//Of for j
		}//Of for i
		
		environment.reset();
	}//Of reset

	/**
	 ****************** 
	 * Train.
	 * 
	 * @param paraEpisodes
	 *            The number of episodes.
	 ****************** 
	 */
	public void train(int paraEpisodes, double paraEpsilon, double paraAlpha) {
		//System.out.println("Training stage ...");
		// Step 1. Initialize.
		environment.reset();

		for (int i = 0; i < agentArray.length; i++) {
			agentArray[i].reset();
			agentArray[i].setTrainingStage(true);
			agentArray[i].setEpsilon(paraEpsilon);
			agentArray[i].setAlpha(paraAlpha);
		}//Of for i

		Arrays.fill(winTimesArray, 0);

		// Step 2. Run the given episodes.
		for (int i = 0; i < paraEpisodes; i++) {
			if ((i == 0) || (i >= paraEpisodes - 5) ) {
				//Important: change to true for tracking.
				SimpleTools.variableTracking = false;
			} else {
				SimpleTools.variableTracking = false;
			}//Of if

			//SimpleTools.variableTrackingOutput("\r\nEpisode " + i + ":\r\n");
			//Step 2.1. Reinitialize the environment while not the players.
			environment.reset();

			int tempCurrentPlayer = 0;
			int tempGameSituation = TicTacToe.UNFINISHED;

			// Step 2.2. Each time a final state should be reached.
			while (tempGameSituation == TicTacToe.UNFINISHED) {
				try {
					tempGameSituation = agentArray[tempCurrentPlayer].step();
				} catch (Exception ee) {
					System.out.println("QAgent for agent.step: " + ee);
					System.exit(0);
				} // Of try

				tempCurrentPlayer = (tempCurrentPlayer + 1) % 2;
			} // Of while

			//SimpleTools.variableTrackingOutput("The environment is: " + environment.toString());
			winTimesArray[tempGameSituation]++;
			SimpleTools.variableTrackingOutput(" & " + tempGameSituation + "\\\\\r\n");
			
			//Step 2.3. Now update the 
			for (int j = 0; j < agentArray.length; j++) {
				agentArray[j].backup();
			}//Of for i
		} // Of for i
		
		//Step 2.3. Only for testing.
		//for (int i = 0; i< agentArray.length; i++) {
		//	agentArray[i].showTheFirstStep();
		//}//Of for i
	} // Of train
	
	/**
	 ****************** 
	 * Play.
	 * 
	 * @param paraEpisodes
	 *            The number of episodes.
	 ****************** 
	 */
	public void play(int paraEpisodes) {
		//System.out.println("Now play ...");
		// Step 1. Initialize.
		for (int i = 0; i < agentArray.length; i++) {
			//agentArray[i].reset();
			agentArray[i].setTrainingStage(false);
		}//Of for i

		Arrays.fill(winTimesArray, 0);

		// Step 2. Run the given episodes.
		for (int i = 0; i < paraEpisodes; i++) {
			//SimpleTools.variableTrackingOutput("\r\nEpisode " + i + ":\r\n");
			//Step 2.1. Reinitialize the environment while not the players.
			environment.reset();

			int tempCurrentPlayer = 0;
			int tempGameSituation = TicTacToe.UNFINISHED;

			// Step 2.2. Each time a final state should be reached.
			while (tempGameSituation == TicTacToe.UNFINISHED) {
				try {
					tempGameSituation = agentArray[tempCurrentPlayer].step();
				} catch (Exception ee) {
					System.out.println("QAgent for agent.step: " + ee);
					System.exit(0);
				} // Of try

				tempCurrentPlayer = (tempCurrentPlayer + 1) % 2;
			} // Of while

			//SimpleTools.variableTrackingOutput("The environment is: " + environment.toString());
			winTimesArray[tempGameSituation]++;
			SimpleTools.variableTrackingOutput(" & " + tempGameSituation + "\\\\");
		} // Of for i
		SimpleTools.variableTrackingOutput("" + environment);
	} // Of play
	
	/**
	 ****************** 
	 * Play with human.
	 * 
	 * @param paraPosition
	 *            The human player's new position.
	 * @return the game situation result.
	 ****************** 
	 */
	public int humanPlay(int paraAction) {
		try {
			environment.step(paraAction);
		} catch (Exception ee) {
			System.out.println("Internal error from VUmpire.step()");
			System.exit(0);
		}//Of try
		
		return environment.getGameSituation();
		//if (environment.getGameSituation() == VTicTacToe.WHITE) {
		//	return VTicTacToe.WHITE;
		//} else if (environment.getGameSituation() == VTicTacToe.BLACK) {
		//	return VTicTacToe.BLACK;
		//}//Of if
		
		//winner = agentArray[1].step();
		
		//result = agentArray[1].getRecentAction();
		//return result;
	}//Of humanPlay

	/**
	 ****************** 
	 * Play with agent.
	 * 
	 * @param paraPosition
	 *            The human player's new position.
	 * @return the result. -1 stands for the human win, 9 stands for the machine win
	 ****************** 
	 */
	public int agentPlay() {
		int result = 0;
		winner = agentArray[1].step();
		return winner;
	}//Of agentPlay	
	
	/**
	 ****************** 
	 * Getter.
	 ****************** 
	 */
	public int getRecentAction() {
		return agentArray[1].getRecentAction();
	}// Of getRecentAction

	/**
	 ****************** 
	 * Getter.
	 * @return The win times array.
	 ****************** 
	 */
	public int[] getWinTimesArray() {
		return winTimesArray;
	}//Of getWinTimesArray
	
	/**
	 ****************** 
	 * Test.
	 ****************** 
	 */
	public static void ticTacToeVTest() {
		VUmpire tempUmpire = new VUmpire();

		int tempEpisodes = 10000;
		SimpleTools.variableTracking = false;
		tempUmpire.train(tempEpisodes, 0.1, 0.1);

		System.out.println("\r\nEpisodes = " + tempEpisodes + ", winTimesArray = "
				+ Arrays.toString(tempUmpire.winTimesArray));

		tempUmpire.play(1);
	}// Of ticTacToeVTest

	/**
	 ****************** 
	 * For unit test.
	 * 
	 * @param args
	 *            Not provided.
	 ****************** 
	 */
	public static void main(String args[]) {
		ticTacToeVTest();
	}// Of main
}// Of class Umpire

File: python/games/vumpire.py
########################################################################
# Copyright (C)                                                        #
# 2020 Fan Min (minfanphd@163.com)                                     #
# 2016 - 2018 Shangtong Zhang (zhangshangtong.cpp@gmail.com)           #
# 2016 Jan Hakenberg (jan.hakenberg@gmail.com)                         #
# 2016 Tian Jun (tianjun.cpp@gmail.com)                                #
# 2016 Kenta Shimada (hyperkentakun@gmail.com)                         #
# Permission given to modify the code as long as you keep this         #
# declaration at the top                                               #
# https://github.com/FanSmale/MFReinforcement/tree/master/python/games #
########################################################################

import numpy as np
import checkerboard
import vagent
import player

############################
# The umpire who is in charge of the whole process.
############################
class VUmpire:
	################
	# The constructor.
	################
	def __init__(self, para_epsilon = 0.1, para_alpha = 0.1):
		# Step 1. Accept parameters.
		self.environment = checkerboard.Checkerboard()
		self.agent_array = []
		self.agent_array.append(vagent.VAgent(self.environment, checkerboard.WHITE, para_epsilon, para_alpha))
		self.agent_array.append(vagent.VAgent(self.environment, checkerboard.BLACK, para_epsilon, para_alpha))

		# Step 2. Initialize member variables.
		self.winner = 0
		self.win_times_array = np.zeros(3)
		self.player_checker_board = np.zeros((checkerboard.SIZE, checkerboard.SIZE))
		
	################
	# Reset for a new game.
	################
	def reset_for_game(self):
		self.winner = checkerboard.EMPTY
		for i in range(checkerboard.SIZE):
			for j in range(checkerboard.SIZE):
				self.player_checker_board[i][j] = checkerboard.EMPTY
		self.environment.reset()

	################
	# Print the learning results.
	################
	def print_learning_results(self):
		print("Agent 1: ")
		self.agent_array[0].print_learning_results()
		print("Agent 2: ")
		self.agent_array[1].print_learning_results()

        ################
	# Train.
	################
	def train(self, para_episodes = 1000):
		self.win_times_array = np.zeros(3)
		for i in range(2):
			self.agent_array[i].reset()
		
		for i in range(para_episodes):
                        # Step 1.1 Reset.
			self.environment.reset()
			temp_current_player = 0;
			temp_game_situation = checkerboard.UNFINISHED

			# Step 1.2. Reach a final state each time.
			while (temp_game_situation == checkerboard.UNFINISHED):
				temp_game_situation = self.agent_array[temp_current_player].step()
				temp_current_player = (temp_current_player + 1) % 2
			# print("Game situation: ", temp_game_situation)
			self.win_times_array[temp_game_situation] = self.win_times_array[temp_game_situation] + 1

			# Step 1.3. Now update the agents
			for j in range(2):
				self.agent_array[j].backup()

	################
	# Play. For agents there is only one game if the greedy approach is employed.
	################
	def play(self):
                #Step 1. Reset for the play stage.
		self.environment.reset()
		self.environment.print_board()
		temp_current_player = 0;
		temp_game_situation = checkerboard.UNFINISHED;
		for i in range(2):
			self.agent_array[i].set_stage(False)
			#self.agent_array[i].reset()

		# Now play.
		while (temp_game_situation == checkerboard.UNFINISHED):
			temp_game_situation = self.agent_array[temp_current_player].step()
			temp_current_player = (temp_current_player + 1) % 2
			self.environment.print_board()
		return temp_game_situation


	################
	# Play. For agents there is only one game if the greedy approach is employed.
	################
	def human_machine_play(self):
                #Step 1. Reset for the play stage.
		self.environment.reset()
		temp_current_player = 0;
		temp_game_situation = checkerboard.UNFINISHED;

		self.agent_array[0] = player.HumanPlayer(self.environment)
		#self.agent_array[1].set_stage(False)

		temp_winner = self.play()
		if (temp_winner == checkerboard.WHITE):
			print("You win!")
		elif (temp_winner == checkerboard.BLACK):
			print("You lose :<")
		else:
			print("It is a tie")
            
################
#The main entrance of the program.
################
if __name__ == '__main__':
	temp_umpire = VUmpire(para_epsilon = 0.01, para_alpha = 0.1)
	temp_rounds = int(1e4)
	print("***Training rounds (change this setting to 1e5 in vumpire.py to obtain a smart agent): ", temp_rounds)
	temp_umpire.train(temp_rounds)
	print("In the training stage, win times (tie, white, black): ", temp_umpire.win_times_array)
	temp_umpire.print_learning_results()

	print("***Now two agents plays one game: ")
	temp_winner = temp_umpire.play()
	print("The winner is (0 for tie, 1 for white and 2 for black): ", temp_winner)
	
	print("***Now you can play with me :)")
	temp_umpire.human_machine_play()
	print("***The end.")

